torch/csrc/jit/passes/utils/subgraph_utils.cpp

```diff
@@ -233,68 +253,77 @@ void mergeNodeIntoSubgraph(
         auto nv = subgraph->insertConstant(*value);
         nv->copyMetadata(input);
-        inputsMap[input] = nv;
+        externalValuesMap[input] = nv;
       } else {
         // The common case: this is a regular input, so just register it with
         // the group node and inner subgraph
         subgraphNode->addInput(input);
         auto inputToGraph = subgraph->addInput();
         inputToGraph->copyMetadata(input);
-        inputsMap[input] = inputToGraph;
+        externalValuesMap[input] = inputToGraph;
       }
     }
   }
 
   // Merge the node into the graph
-  auto mergedNode = subgraph->insertNode(
-      subgraph->createClone(toMerge, [&](Value* v) { return inputsMap[v]; }));
-
-  // If n's outputs were inputs to `group`, remove them since we just merged
-  // n in.
-  //
-  // i.e.,
-  // x = f(w); group(x, y, z) becomes group(w, y, z).
-  // x, y, z = f(w); group(x, y, z) becomes group(w).
-  auto inputs = subgraphNode->inputs();
-  for (size_t i = 0; i < toMerge->outputs().size(); ++i) {
-    auto it = std::find(inputs.begin(), inputs.end(), toMerge->outputs()[i]);
-    if (it != inputs.end()) {
-      size_t p = it - inputs.begin();
-      subgraphNode->removeInput(p);
-      subgraph->inputs()[p]->replaceAllUsesWith(mergedNode->outputs()[i]);
-      subgraph->eraseInput(p);
+  auto mergedNode = subgraph->insertNode(subgraph->createClone(
+      toMerge, [&](Value* v) { return externalValuesMap[v]; }));
+
+  if (!merging_node_after_subgraph) {
+    // If n's outputs were inputs to `group`, remove them since we just merged
+    // n in.
+    //
+    // i.e.,
+    // x = f(w); group(x, y, z) becomes group(w, y, z).
+    // x, y, z = f(w); group(x, y, z) becomes group(w).
+    auto inputs = subgraphNode->inputs();
+    for (size_t i = 0; i < toMerge->outputs().size(); ++i) {
+      auto it = std::find(inputs.begin(), inputs.end(), toMerge->outputs()[i]);
+      if (it != inputs.end()) {
+        size_t p = it - inputs.begin();
+        subgraphNode->removeInput(p);
+        subgraph->inputs()[p]->replaceAllUsesWith(mergedNode->outputs()[i]);
+        subgraph->eraseInput(p);
+      }
     }
   }
 
   // Add n's outputs to the group node and inner subgraph outputs.
   for (size_t i = 0; i < toMerge->outputs().size(); i++) {
     auto oldOutput = toMerge->outputs()[i];
-
-    // Only register the output in the group node if it's actually used
-    // outside the subgraph.
-    const auto hasUsesOutsideSubgraph = std::any_of(
-        oldOutput->uses().cbegin(),
-        oldOutput->uses().cend(),
-        [&](const Use& use) { return use.user->isAfter(subgraphNode); });
-
-    if (hasUsesOutsideSubgraph) {
-      auto newOutput = mergedNode->outputs()[i];
-      subgraph->registerOutput(newOutput);
-      auto groupOutput = subgraphNode->addOutput();
-      groupOutput->copyMetadata(oldOutput);
-      oldOutput->replaceAllUsesWith(groupOutput);
-    }
+    auto newOutput = mergedNode->outputs()[i];
```

Comment:
why did we hoist pruning unused outputs outside and after this loop?

Reply:
Because we want to do it after we destroy the node

Reply:
Might be worth adding a comment saying exactly that :)